### ISSUE TITLE
Master<->Slave sync with >2GB RDB file

### ISSUE THREAD

**Issue by svantreeck** [microsoftarchive/redis#166]
We currently have a redis database that is about 20M keys that takes about 5GB memory/3GB RDB file.  We have attempted to add slaves to it, and when the slave starts up, it seems like it attempts a full sync, but only tries to transfer 2GB of data, fails because the data it has received is "corrupt" (I think just missing about 1gb of the transfer), and then shuts down.

I tried checking the config file if there was anything we could manipulate to change the amount of data that could be transfered during a full sync; however, I either missed the setting or cannot find what is going on.

Below is the log from the slave as it attempts to startup:

```
[4564] 04 Sep 12:09:02.640 # Server started, Redis version 2.8.13
[4564] 04 Sep 12:09:02.640 * The server is now ready to accept connections on port 6380
[4564] 04 Sep 12:09:02.655 * Connecting to MASTER 127.0.0.1:6379
[4564] 04 Sep 12:09:02.655 * MASTER <-> SLAVE sync started
[4564] 04 Sep 12:09:02.655 * Non blocking connect for SYNC fired the event.
[4564] 04 Sep 12:09:02.655 * Master replied to PING, replication can continue...
[4564] 04 Sep 12:09:02.655 * Partial resynchronization not possible (no cached master)
[4564] 04 Sep 12:09:02.655 * Full resync from master: 9b5c5f7ad8c3915a069fbba2f6e9ee031f24a617:5346
[4564] 04 Sep 12:10:09.378 * MASTER <-> SLAVE sync: receiving 2147483647 bytes from master
[4564] 04 Sep 12:14:23.788 * MASTER <-> SLAVE sync: Flushing old data
[4564] 04 Sep 12:14:23.788 * MASTER <-> SLAVE sync: Loading DB in memory
[4564] 04 Sep 12:15:15.643 # Short read or OOM loading DB. Unrecoverable error, aborting now.
```

Let me know if there is additional information that would be helpful in getting the resync to work.

**Comment by jepickett** [comment]
2147483647 = 0x7FFFFFFF. This is probably a data type overflow.

**Comment by svantreeck** [comment]
Yeah, the size is suspicious.  Is this something that can be fixed in this version of redis?  I am running on a 64 bit server, so ideally, I'd think it the size could be expanded to a long, but I'm not sure what repercussions that would involve.

**Comment by jepickett** [comment]
I have a fix for this in commit 874c21d801ad4fc43fd2f480cfc7dc50f697b49d. Let me know if this resolves the problem on your machine.

**Comment by svantreeck** [comment]
works great, thanks for the help!
